### PR TITLE
Pipelines for 3.4

### PR DESCRIPTION
**Pull request by ehelms**

### CODE REVIEW

File: pipelines/pipeline_katello_34.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        --foreman-proxy-content-certs-tar "{{ foreman_proxy_content_certs_tar }}"
+        --foreman-proxy-content-parent-fqdn "{{ server_fqdn.stdout }}"
+        --foreman-proxy-content-pulp-oauth-secret "{{ pulp_oauth_secret.stdout }}"'
+    foreman_installer_additional_packages:
+      - foreman-installer-katello
+  roles:
+    - foreman_proxy_content
+    - foreman_installer
+
+- hosts: pipeline-proxy-3.4-centos7
```
Comment (evgeni) [inline pipelines/pipeline_katello_34.yml]: no bats on the katello box?

File: pipelines/pipeline_katello_32_to_34_upgrade.yml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    katello_repositories_use_koji: True
+    foreman_installer_scenario: katello
+    foreman_installer_options_internal_use_only:
+        - "{{ '--disable-system-checks' if katello_repositories_version != '3.2' else '' }}"
+  roles:
+    - foreman_repositories
+    - katello_repositories
+    - foreman_installer
+
+- hosts: pipeline-proxy-3.2-3.4-centos7
```
Comment (evgeni) [inline pipelines/pipeline_katello_32_to_34_upgrade.yml]: no bats on the katello box after the upgrade?

File: pipelines/pipeline_katello_33_to_34_upgrade.yml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    foreman_installer_scenario: katello
+    foreman_installer_options_internal_use_only:
+        - "{{ '--disable-system-checks' if katello_repositories_version != '3.3' else '' }}"
+  roles:
+    - puppet_repositories
+    - foreman_repositories
+    - katello_repositories
+    - foreman_installer
+
+- hosts: pipeline-proxy-3.3-3.4-centos7
```
Comment (evgeni) [inline pipelines/pipeline_katello_33_to_34_upgrade.yml]: no bats on the katello box after the upgrade?

File: pipelines/pipeline_katello_33_to_34_upgrade.yml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      - foreman-installer-katello
+  roles:
+    - foreman_proxy_content
+    - foreman_installer
+
+- hosts: pipeline-katello-3.3-3.4-centos7
+  become: true
+  vars:
+    bats_tests:
+      - "fb-content-katello.bats"
```
Comment (evgeni) [inline pipelines/pipeline_katello_33_to_34_upgrade.yml]: no swapfile.bats here? the other pipelines have it

File: pipelines/pipeline_katello_32_to_34_upgrade.yml (hunk 2 above)
Comment (ehelms) [inline pipelines/pipeline_katello_32_to_34_upgrade.yml]: Nope - cause bats isn't designed around being re-run at the moment, see #422

File: pipelines/pipeline_katello_33_to_34_upgrade.yml (hunk 4 above)
Comment (ehelms) [inline pipelines/pipeline_katello_33_to_34_upgrade.yml]: So far I have not seen a need for it. IIRC it was originally added due to limitations and timeouts on our CI system.

File: pipelines/pipeline_katello_33_to_34_upgrade.yml (hunk 3 above)
Comment (ehelms) [inline pipelines/pipeline_katello_33_to_34_upgrade.yml]: Nope - cause bats isn't designed around being re-run at the moment, see #422

File: pipelines/pipeline_katello_34.yml (hunk 1 above)
Comment (ehelms) [inline pipelines/pipeline_katello_34.yml]: Nope - cause bats isn't designed around being re-run at the moment, see #422

File: pipelines/pipeline_katello_33_to_34_upgrade.yml (hunk 4 above)
Comment (evgeni) [inline pipelines/pipeline_katello_33_to_34_upgrade.yml]: ACK, just wanted to be sure.

File: pipelines/pipeline_katello_34.yml (hunk 1 above)
Comment (evgeni) [inline pipelines/pipeline_katello_34.yml]: (Just tested that, and adding bats to the katello box here works fine)

Comment (ehelms) [inline pipelines/pipeline_katello_34.yml]: What I mean is, I cannot run bats here, upgrade, then run bats again (without deleting the data before upgrade)

**Comment by ehelms** [comment]
@evgeni What else would you like to see before this gets merged in? I keep forgetting I use this locally and its not available upstream.

**Comment by evgeni** [comment]
Otherwise I think I'm good.

**Comment by evgeni** [comment]
Ah, just looked at my local changes, you want:
```
+    - disable_firewall
+    - haveged
```
in every playbook after the `*_repositories` roles.

**Comment by ehelms** [comment]
Why do we need to disable the firewall and enabled haveged for these?

**Comment by evgeni** [comment]
For consistency? Having haveged on a VM is never a bad idea (don't quote me on that) and disable firewall should be a noop in most cases (but is enabled in the other pipeline playbooks we have)

**Comment by ehelms** [comment]
I've updated this based on comments and added bats testing into the pipelines. This is ready for re-review.

File: pipelines/pipeline_katello_32_to_34_upgrade.yml [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  become: yes
+  vars:
+    foreman_proxy_content_server: pipeline-katello-3.2-3.4-centos7
+    foreman_installer_scenario: capsule
+    foreman_installer_options_internal_use_only:
+      - "--disable-system-checks"
+      - '--foreman-proxy-trusted-hosts "{{ server_fqdn.stdout }}"
```
Comment (ekohl) [inline pipelines/pipeline_katello_32_to_34_upgrade.yml]: Why pass this in as a long string? We can pass in an array of options.

File: pipelines/pipeline_katello_32_to_34_upgrade.yml [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      - katello
+  roles:
+    - foreman_installer
+
+- hosts: pipeline-proxy-3.2-3.4-centos7
+  become: yes
+  vars:
+    foreman_proxy_content_server: pipeline-katello-3.2-3.4-centos7
+    foreman_installer_scenario: capsule
+    foreman_installer_options_internal_use_only:
```
Comment (ekohl) [inline pipelines/pipeline_katello_32_to_34_upgrade.yml]: We should really fix the internal_use_only vs plain options sometime.

File: pipelines/pipeline_katello_32_to_34_upgrade.yml [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  become: true
+  vars:
+    katello_repositories_version: 3.4
+    foreman_repositories_version: 1.15
+    foreman_installer_upgrade: True
+    foreman_repositories_use_koji: True
+    katello_repositories_use_koji: True
+    foreman_installer_scenario: katello
+    foreman_installer_options_internal_use_only:
+        - "{{ '--disable-system-checks' if katello_repositories_version != '3.2' else '' }}"
```
Comment (ekohl) [inline pipelines/pipeline_katello_32_to_34_upgrade.yml]: Since we also set the variable here and know it's 3.4, isn't it always true?

File: pipelines/pipeline_katello_33_to_34_upgrade.yml [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    - hammer_config
+
+- hosts: pipeline-proxy-3.3-3.4-centos7
+  become: yes
+  vars:
+    foreman_proxy_content_server: pipeline-katello-3.3-3.4-centos7
+    foreman_installer_scenario: foreman-proxy-content
+    foreman_installer_options_internal_use_only:
+      - "--disable-system-checks"
+      - '--foreman-proxy-trusted-hosts "{{ server_fqdn.stdout }}"
```
Comment (ekohl) [inline pipelines/pipeline_katello_33_to_34_upgrade.yml]: Can also be an array

File: pipelines/pipeline_katello_33_to_34_upgrade.yml [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  become: true
+  vars:
+    katello_repositories_version: 3.4
+    foreman_repositories_version: 1.15
+    foreman_installer_upgrade: True
+    foreman_repositories_use_koji: True
+    katello_repositories_use_koji: True
+    foreman_installer_scenario: katello
+    foreman_installer_options_internal_use_only:
+        - "{{ '--disable-system-checks' if katello_repositories_version != '3.3' else '' }}"
```
Comment (ekohl) [inline pipelines/pipeline_katello_33_to_34_upgrade.yml]: I believe this is also always true.

File: pipelines/pipeline_katello_33_to_nightly_upgrade.yml [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    puppet_repositories_version: 4
+    foreman_proxy_content_server: pipeline-katello-3.3-nightly-centos7
+    katello_repositories_version: 3.3
+    foreman_repositories_version: 1.14
+    foreman_repositories_use_koji: true
+    katello_repositories_use_koji: true
+    foreman_installer_scenario: foreman-proxy-content
+    foreman_installer_options_internal_use_only:
+      - "--disable-system-checks"
+      - '--foreman-proxy-trusted-hosts "{{ server_fqdn.stdout }}"
```
Comment (ekohl) [inline pipelines/pipeline_katello_33_to_nightly_upgrade.yml]: Can also be an array?

File: pipelines/pipeline_katello_33_to_nightly_upgrade.yml [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+  become: true
+  vars:
+    katello_repositories_version: nightly
+    foreman_repositories_version: nightly
+    foreman_installer_upgrade: True
+    foreman_repositories_use_koji: True
+    katello_repositories_use_koji: True
+    foreman_installer_scenario: katello
+    foreman_installer_options_internal_use_only:
+        - "{{ '--disable-system-checks' if katello_repositories_version != '3.3' else '' }}"
```
Comment (ekohl) [inline pipelines/pipeline_katello_33_to_nightly_upgrade.yml]: Also always true?

File: pipelines/pipeline_katello_33_to_34_upgrade.yml [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+  vars:
+    foreman_installer_scenario: katello
+    foreman_installer_options_internal_use_only:
+      - "--foreman-admin-password {{ foreman_installer_admin_password }}"
+      - "--disable-system-checks"
+    foreman_installer_additional_packages:
+      - katello
+  roles:
+    - foreman_installer
+    - hammer_config
```
Comment (ekohl) [inline pipelines/pipeline_katello_33_to_34_upgrade.yml]: In plain foreman there's an option to enable the CLI which should set it up for you. Won't that be sufficient or is it not available in the katello scenario?

File: pipelines/pipeline_katello_34.yml [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  become: yes
+  vars:
+    foreman_proxy_content_server: pipeline-katello-3.4-centos7
+    foreman_installer_scenario: foreman-proxy-content
+    foreman_installer_options_internal_use_only:
+      - '--disable-system-checks
+        --foreman-proxy-trusted-hosts "{{ server_fqdn.stdout }}"
```
Comment (ekohl) [inline pipelines/pipeline_katello_34.yml]: Can also be an array

File: pipelines/pipeline_katello_32_to_34_upgrade.yml (hunk 5 above)
Comment (ehelms) [inline pipelines/pipeline_katello_32_to_34_upgrade.yml]: To the installer option? How so?

File: pipelines/pipeline_katello_33_to_34_upgrade.yml (hunk 12 above)
Comment (ehelms) [inline pipelines/pipeline_katello_33_to_34_upgrade.yml]: Not available in Katello 3.3, so this has to be configured to run the bats test before upgrading :/

File: pipelines/pipeline_katello_32_to_34_upgrade.yml (hunk 5 above)
Comment (ekohl) [inline pipelines/pipeline_katello_32_to_34_upgrade.yml]: Doesn't this work:
```yaml
foreman_installer_options_internal_use_only:
  - "--disable-system-checks"
  - '--foreman-proxy-trusted-hosts "{{ server_fqdn.stdout }}"'
  - '--foreman-proxy-trusted-hosts "{{ ansible_nodename }}"'
  - '--foreman-proxy-foreman-base-url "https://{{ server_fqdn.stdout }}"'
  - '--foreman-proxy-register-in-foreman true'
  - '--foreman-proxy-oauth-consumer-key "{{ oauth_consumer_key.stdout }}"'
  - '--foreman-proxy-oauth-consumer-secret "{{ oauth_consumer_secret.stdout }}"'
  - '--capsule-certs-tar "{{ foreman_proxy_content_certs_tar }}"'
  - '--capsule-parent-fqdn "{{ server_fqdn.stdout }}"'
  - '--capsule-pulp-oauth-secret "{{ pulp_oauth_secret.stdout }}"''
```

File: pipelines/pipeline_katello_34.yml [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    bats_tests:
+      - fb-test-katello.bats
+      - fb-content-katello.bats
+      - fb-proxy.bats
+      - fb-destroy-organization.bats
+      - fb-finish-katello.bats
+  roles:
+    - bats
+
+- hosts: pipeline-proxy-3.4-centos7
```
Comment (evgeni) [inline pipelines/pipeline_katello_34.yml]: this can go, we don't (and currently can't) run proxy.bats on the proxy.

File: pipelines/pipeline_katello_32_to_34_upgrade.yml [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+      - foreman-installer-katello
+  roles:
+    - foreman_proxy_content
+    - foreman_installer
+
+- hosts: pipeline-katello-3.2-3.4-centos7
+  become: true
+  vars:
+    bats_tests:
+      - "swapfile.bats"
```
Comment (evgeni) [inline pipelines/pipeline_katello_32_to_34_upgrade.yml]: `swapfile.bats` is no more

File: pipelines/pipeline_katello_33_to_34_upgrade.yml [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    bats_tests:
+      - fb-test-katello.bats
+      - fb-destroy-organization.bats
+      - fb-content-katello.bats
+      - fb-proxy.bats
+      - fb-finish-katello.bats
+  roles:
+    - bats
+
+- hosts: pipeline-proxy-3.3-3.4-centos7
```
Comment (evgeni) [inline pipelines/pipeline_katello_33_to_34_upgrade.yml]: no proxy.bats on the proxy

File: pipelines/pipeline_katello_33_to_nightly_upgrade.yml [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+    - katello_repositories
+    - foreman_proxy_content
+    - foreman_installer
+
+- hosts: pipeline-katello-3.3-nightly-centos7
+  become: true
+  vars:
+    bats_update_forklift: no
+    bats_tests:
+      - "swapfile.bats"
```
Comment (evgeni) [inline pipelines/pipeline_katello_33_to_nightly_upgrade.yml]: `swapfile.bats` is no more

File: pipelines/pipeline_katello_33_to_nightly_upgrade.yml [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+  vars:
+    bats_forklift_dir: "/home/vagrant/sync"
+    bats_update_forklift: no
+    bats_tests:
+      - "fb-content-upgrade.bats"
+      - "fb-finish-katello.bats"
+  roles:
+    - bats
+
+- hosts: pipeline-proxy-3.3-nightly-centos7
```
Comment (evgeni) [inline pipelines/pipeline_katello_33_to_nightly_upgrade.yml]: no proxy.bats on the proxy, can be done on the katello host though

File: pipelines/pipeline_katello_32_to_34_upgrade.yml (hunk 5 above)
Comment (evgeni) [inline pipelines/pipeline_katello_32_to_34_upgrade.yml]: I think like this:

```yaml
   foreman_installer_options_internal_use_only:
     - "--disable-system-checks"
     - '--foreman-proxy-trusted-hosts "{{ server_fqdn.stdout }}"'
     - '--foreman-proxy-trusted-hosts "{{ ansible_nodename }}"'
     - …
```

**Comment by ehelms** [comment]
@evgeni @ekohl OK - I think I addressed all the issues now.

File: pipelines/pipeline_katello_33_to_nightly_upgrade.yml [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+  roles:
+    - puppet_repositories
+    - foreman_repositories
+    - katello_repositories
+    - foreman_installer
+
+- hosts: pipeline-katello-3.3-nightly-centos7
+  become: true
+  vars:
+    bats_forklift_dir: "/home/vagrant/sync"
```
Comment (ekohl) [inline pipelines/pipeline_katello_33_to_nightly_upgrade.yml]: Other pipelines don't use this. I'd like it to be consistent.

File: pipelines/pipeline_katello_32_to_34_upgrade.yml [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+    - update_os_packages
+    - foreman_installer
+
+- hosts: pipeline-katello-3.2-3.4-centos7
+  become: yes
+  vars:
+    foreman_installer_scenario: katello
+    foreman_installer_options_internal_use_only:
+      - "--foreman-admin-password {{ foreman_installer_admin_password }}"
+      - "{{ '--disable-system-checks' if katello_repositories_version != '3.2' else '' }}"
```
Comment (evgeni) [inline pipelines/pipeline_katello_32_to_34_upgrade.yml]: Right, it is only defined in the step before.

File: pipelines/pipeline_katello_33_to_nightly_upgrade.yml [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+    bats_update_forklift: no
+    bats_tests:
+      - "fb-content-upgrade.bats"
```
Comment (evgeni) [inline pipelines/pipeline_katello_33_to_nightly_upgrade.yml]: we don't need quotes here (and we don't have them in other places)

File: pipelines/pipeline_katello_33_to_nightly_upgrade.yml [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+    - katello_repositories
+    - foreman_proxy_content
+    - foreman_installer
+
+- hosts: pipeline-katello-3.3-nightly-centos7
+  become: true
+  vars:
+    bats_update_forklift: no
+    bats_tests:
+      - "fb-content-katello.bats"
```
Comment (evgeni) [inline pipelines/pipeline_katello_33_to_nightly_upgrade.yml]: we don't need quotes here (and we don't have them in other places)

File: pipelines/pipeline_katello_32_to_34_upgrade.yml [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+      - foreman-installer-katello
+  roles:
+    - foreman_proxy_content
+    - foreman_installer
+
+- hosts: pipeline-katello-3.2-3.4-centos7
+  become: true
+  vars:
+    bats_tests:
+      - "fb-content-katello.bats"
```
Comment (evgeni) [inline pipelines/pipeline_katello_32_to_34_upgrade.yml]: we don't need quotes here (and we don't have them in other places)

**Comment by ekohl** [comment]
Now that we're adding 3.5, we should really get this merged.

File: pipelines/pipeline_katello_34.yml [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+    katello_repositories_use_koji: true
+    foreman_installer_skip_installer: true
+  roles:
+    - selinux
+    - etc_hosts
+    - epel_repositories
+    - puppet_repositories
+    - foreman_repositories
+    - katello_repositories
+    - pulp_repositories
```
Comment (ekohl) [inline pipelines/pipeline_katello_34.yml]: This role doesn't exist

**Comment by ehelms** [comment]
@evgeni @ekohl updated

File: pipelines/pipeline_katello_nightly.yml [hunk 25]
```diff
@@ -69,6 +69,12 @@
 
 - hosts: pipeline-katello-nightly-centos7
   become: true
+  tasks:
```
Comment (ekohl) [inline pipelines/pipeline_katello_nightly.yml]: Why is this needed now? Is it an additional test so we spot issues like the puppetserver one? I'd also like a reboot test but that may be trickier to implement.

**Comment by ekohl** [comment]
I do get a pipeline failure when I run the plain 3.4 (no upgrade) pipeline.
```
not ok 3 sync proxy
# (in test file fb-proxy.bats, line 26)
#   `hammer capsule content synchronize --id=$PROXY_ID' failed with status 70
# Could not synchronize capsule content:
#   Error: Request Timeout
not ok 4 content is available from proxy
# (in test file fb-proxy.bats, line 32)
#   `wget $URL1 -P /tmp || wget $URL2 -P /tmp' failed with status 8
# --2017-10-13 08:36:36--  http://pipeline-proxy-3-4-centos7.lab.kohlvanwijngaarden.nl/pulp/repos/Test_Organization/Library/Test_CV/custom/Test_Product/Zoo/walrus-0.71-1.noarch.rpm
# Resolving pipeline-proxy-3-4-centos7.lab.kohlvanwijngaarden.nl (pipeline-proxy-3-4-centos7.lab.kohlvanwijngaarden.nl)... 192.168.121.90, 192.168.121.90
# Connecting to pipeline-proxy-3-4-centos7.lab.kohlvanwijngaarden.nl (pipeline-proxy-3-4-centos7.lab.kohlvanwijngaarden.nl)|192.168.121.90|:80... connected.
# HTTP request sent, awaiting response... 404 NOT FOUND
# 2017-10-13 08:36:36 ERROR 404: NOT FOUND.
# 
# --2017-10-13 08:36:36--  http://pipeline-proxy-3-4-centos7.lab.kohlvanwijngaarden.nl/pulp/repos/Test_Organization/Library/Test_CV/custom/Test_Product/Zoo/Packages/w/walrus-0.71-1.noarch.rpm
# Resolving pipeline-proxy-3-4-centos7.lab.kohlvanwijngaarden.nl (pipeline-proxy-3-4-centos7.lab.kohlvanwijngaarden.nl)... 192.168.121.90, 192.168.121.90
# Connecting to pipeline-proxy-3-4-centos7.lab.kohlvanwijngaarden.nl (pipeline-proxy-3-4-centos7.lab.kohlvanwijngaarden.nl)|192.168.121.90|:80... connected.
# HTTP request sent, awaiting response... 404 NOT FOUND
# 2017-10-13 08:36:36 ERROR 404: NOT FOUND.
```

**Comment by evgeni** [comment]
Mhh, I get the same timeout :(

**Comment by ehelms** [comment]
@evgeni @ekohl I am not sure what y;all want me to do with it at this point given the timeout you noted in the tests. That seems less an issue with the pipeline as it does possibly the code.

**Comment by ekohl** [comment]
@ehelms let's merge the pipeline and deal with it as a general problem. You might be right that it's in the code because we see it in all versions.

**Comment by evgeni** [comment]
Yeah, I agree, the pipeline looks fine from a code pov, the problem is somewhere else.
